### ISSUE TITLE
RStudio crashes with XGboost multiclass classification

### ISSUE THREAD

**Issue by RanaivosonHerimanitra** [tqchen/xgboost#153]
Hi there ,
I wrote a high level function that facilates my life when training Xgboost model:

```
train_xgboost= function(validation=NULL,
                        training=NULL,
                        subsample=1,
                        nrounds=50,
                        trees_depth=7,
                        eta=0.2,
                        missing=median,
                        myobjective="multi:softmax",
                        directory=getwd(),
                        features.pattern=NULL,outputlabel=NULL,target="label") 
{
  #data preparation before training:
  data=as.data.frame(training)
  data=data[,c(target,grep(features.pattern,names(data),value=T))]
  trainingdata <- xgb.DMatrix(as.matrix(data[,-1]), label = data[,target],missing=missing)
  mynumclass=length(unique(data[,target]))
  rm(data);gc()
  #validation data preparation:
  validation=as.data.frame(validation)
  validation=validation[,c(target,grep(features.pattern,names(validation),value=T))]
  validation <- xgb.DMatrix(as.matrix(validation[,-1]), label = validation[,target],missing=missing)
  
  #set parameters:
  param <- list("objective" =myobjective ,"bst:eta" = eta,"bst:subsample"=subsample,
                "bst:max_depth" = trees_depth,"eval_metric"="auc",
                "silent" = 0,"nthread" = 7,"num_class"=mynumclass)
  # training begins here:
  mywatchlist <- list(eval = validation, train = eval(parse(text=paste0('trainingdata'))) )
  xgbmodel=xgb.train( params=param, data= eval(parse(text=paste0('trainingdata'))), nrounds=nrounds, verbose=2,watchlist=mywatchlist ) 
  gc();
  #save the model:
  rm(trainingdata);gc()
  xgb.save(xgbmodel,paste0(directory,"_modelwith_",outputlabel))
}
```
It works pretty fine when `myobjective` is set to "binary:logistic"  but when I switch to "multi:softmax",  RStudio crashes(or I have already set the number of class).

Is it a bug? How could I fix it.
Thanks,

**Comment by RanaivosonHerimanitra** [comment]
Here is the data to reproduce the crash : https://github.com/RanaivosonHerimanitra/Xgboost-multiclass-classification using the following command:
```
require(data.table)
mydata=fread('mydata.csv') #set up your path
set.seed(956)
validation_range=sample(1:nrow(mydata),30,replace=F)
training_range=which( !(1:nrow(mydata) %in% validation_range ))
train_xgboost(validation=mydata[validation_range,],
                  training=mydata[training_range,],
                  nrounds=125, 
                  colsample_bytree=0.5,
                  missing=median,
                  trees_depth=7,eta=0.01,
                  subsample=0.5,
                  features.pattern="^x_",
                  outputlabel="xgboost",target="mytarget")
```

**Comment by tqchen** [comment]
Currently multi-class classification accepts input in [0, num_class). You will need to remap your target into this range.

**Comment by RanaivosonHerimanitra** [comment]
Ok, I will rerun and let you know before closing

**Comment by hetong007** [comment]
Hi @RanaivosonHerimanitra , it is not crashing with this piece.

```R
train_xgboost= function(validation=NULL,
                        training=NULL,
                        subsample=1,
                        nrounds=50,
                        trees_depth=7,
                        eta=0.2,
                        missing=median,
                        myobjective="multi:softmax",
                        directory=getwd(),
                        features.pattern=NULL,outputlabel=NULL,target="label") 
{
    #data preparation before training:
    data=as.data.frame(training)
    data=data[,c(target,grep(features.pattern,names(data),value=T))]
    lbs = sort(unique(data[,target]))
    lbs = match(data[,target],lbs)-1
    trainingdata <- xgb.DMatrix(as.matrix(data[,-1]), label = lbs,missing=missing)
    mynumclass=length(unique(data[,target]))
    rm(data);gc()
    #validation data preparation:
    validation=as.data.frame(validation)
    validation=validation[,c(target,grep(features.pattern,names(validation),value=T))]
    validation <- xgb.DMatrix(as.matrix(validation[,-1]), label = validation[,target],missing=missing)
    
    #set parameters:
    param <- list("objective" =myobjective ,"bst:eta" = eta,"bst:subsample"=subsample,
                  "bst:max_depth" = trees_depth,"eval_metric"="auc",
                  "silent" = 0,"nthread" = 7,"num_class"=mynumclass)
    # training begins here:
    mywatchlist <- list(eval = validation, train = eval(parse(text=paste0('trainingdata'))) )
    xgbmodel=xgb.train( params=param, data= eval(parse(text=paste0('trainingdata'))), nrounds=nrounds, verbose=2,watchlist=mywatchlist ) 
    gc();
    #save the model:
    rm(trainingdata);gc()
    xgb.save(xgbmodel,paste0(directory,"_modelwith_",outputlabel))
}
```
I use `lbs` to store transformed labels. It is just a minimal-effort fix. There could be errors waiting for further refinements.

**Comment by RanaivosonHerimanitra** [comment]
lbs is not in  range [0, num_class)

**Comment by RanaivosonHerimanitra** [comment]
I add these code for the validation set :
```
lbs1 = sort(unique(validation[,target]))
 lbs1 = match(validation[,target],lbs1)-1
```
First notice, there is no information about the metric during the model training.
Second,at the end of the training, I got this message:
```
Erreur dans xgb.iter.eval(bst, watchlist, i - 1, feval) : 
  AUC: the dataset only contains pos or neg samples
```

**Comment by hetong007** [comment]
Probably the error is caused by the non-existance of some labels in your validation set. Besides, @tqchen , I am not sure whether the default AUC evaluation could be applied to multiclass jobs. 

If not, @RanaivosonHerimanitra you could try to write your __own measure__ as in [this demo](https://github.com/tqchen/xgboost/blob/master/R-package/demo/custom_objective.R).

**Comment by tqchen** [comment]
AUC is only for binary classification and ranking. For multicass, use merror

On Monday, January 26, 2015, Tong He <notifications@github.com> wrote:

> Probably the error is caused by the non-existance of some labels in your
> validation set. Besides, @tqchen <https://github.com/tqchen> , I am not
> sure whether the default AUC evaluation could be applied to multiclass
> jobs.
>
> If not, @RanaivosonHerimanitra <https://github.com/RanaivosonHerimanitra>
> you could try to write your *own measure* as in this demo
> <https://github.com/tqchen/xgboost/blob/master/R-package/demo/custom_objective.R>
> .
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tqchen/xgboost/issues/153#issuecomment-71536609>.
>


-- 
Sincerely,
------------------------------------------------
Tianqi Chen
Computer Science & Engineering, University of Washington

**Comment by RanaivosonHerimanitra** [comment]
Makes sense!. But in anycase, It would be nice to have an error message instead of a crash.

**Comment by tqchen** [comment]
I have fixed code for more strict boundary check, you should be able to get an error for the initial crash case.

Thanks for the suggestion
